planner/core/task.go

```diff
@@ -1061,28 +1066,61 @@ func (p *PhysicalTopN) pushTopNDownToDynamicPartition(copTsk *copTask) (task, bo
 		tblScan = finalTblScanPlan.(*PhysicalTableScan)
 		tblInfo = tblScan.Table
 	}
+	if len(copTsk.idxMergePartPlans) > 0 {
+		partialFinalScan = make([]PhysicalPlan, 0, len(copTsk.idxMergePartPlans))
+		for _, scan := range copTsk.idxMergePartPlans {
+			clonedScan, err := scan.Clone()
+			if err != nil {
+				return nil, false
+			}
+			partialClonedScan = append(partialClonedScan, clonedScan)
+			finalScan := clonedScan
+			for len(finalScan.Children()) > 0 {
+				finalScan = finalScan.Children()[0]
+			}
+			partialFinalScan = append(partialFinalScan, finalScan)
+		}
+	}
 
 	pi := tblInfo.GetPartitionInfo()
-	if pi == nil {
+	if pi == nil && len(copTsk.idxMergePartPlans) == 0 {
```

Comment:
This change makes the code quite misleading. This method now handles not only partitioned tables but also normal tables. We need to add some comments on this method.

Reply:
I'll change the method's name and add some new comments.